packages/engine/src/worker/mod.rs

```diff
@@ -394,32 +385,54 @@ impl WorkerController {
         Ok(())
     }
 
-    /// Sends a message to runners to cancel the current task and sends `message` to the worker
-    /// pool.
-    async fn finish_task(
+    /// Handles the terminating message of a sub-Task associated with a group (i.e. the last one in
+    /// its execution chain, which will be sent to "Main")
+    ///
+    /// - Drops the `TaskSharedStore` associated with the sub-Task.
+    /// - Removes the `PendingGroup` from the `PendingWorkerTask`, and if there are no more
+    /// `PendingGroup`s then the Task has finished and sends a message to runners to cancel any
+    /// work they're doing on the Task and sends `message` to the worker pool.
+    async fn handle_end_message(
         &mut self,
         task_id: TaskId,
         sim_id: SimulationShortId,
+        group_index: Option<usize>,
         source: Language,
         message: TaskMessage,
         shared_store: TaskSharedStore,
     ) -> Result<()> {
         // `shared_store` metaversioning should have been kept updated
         // by the runners, so it doesn't need to be updated at this point.
-        if let Some(_task) = self.tasks.inner.remove(&task_id) {
-            // Important to drop here since we then lose the access to the shared store
-            drop(shared_store);
-
-            tracing::trace!("Cancelling tasks on the other runners");
-            self.cancel_task_except_for_runner(task_id, source).await?;
-
-            self.worker_pool_comms.send(
-                sim_id,
-                WorkerToWorkerPoolMsg::TaskResultOrCancelled(WorkerTaskResultOrCancelled {
-                    task_id,
-                    payload: TaskResultOrCancelled::Result(message),
-                }),
-            )?;
+        // Important to drop here since we then lose the access to the shared store
+        drop(shared_store);
+
+        if let Entry::Occupied(mut entry) = self.tasks.inner.entry(task_id) {
+            let task = entry.get_mut();
+
+            if let Some(completed_group_index) = task
+                .pending_groups
+                .iter()
+                .position(|group| group.group_index == group_index)
+            {
+                task.pending_groups.remove(completed_group_index);
+            };
+
+            if task.pending_groups.is_empty() {
+                entry.remove();
+
+                tracing::trace!(
+                    "No more pending groups on task [{task_id}], cancelling task on the other \
+                     runners"
+                );
+                self.cancel_task_except_for_runner(task_id, source).await?;
+                self.worker_pool_comms.send(
+                    sim_id,
+                    WorkerToWorkerPoolMsg::TaskResultOrCancelled(WorkerTaskResultOrCancelled {
+                        task_id,
+                        payload: TaskResultOrCancelled::Result(message),
```

Comment:
Actually I'm not sure this is adequate. We're only sending the message of the final group, ignoring the terminating messages of all of the groups that finished first.

@litvand how do we think we want to handle this? It kind of feels like we need another `combine_messages` but on the `WorkerHandler` trait not just the WorkerPoolHandler 🙃 

We could also change `TaskResultOrCancelled` to have a vec of messages I guess?

Reply:
I have done something like that now